core/src/test/java/org/elasticsearch/common/xcontent/ObjectSpoolerTests.java

```diff
@@ -0,0 +1,271 @@
+        e = e.getCause();
+        assertThat(e, instanceOf(IllegalArgumentException.class));
+        assertEquals("Can't repeat param [vegetable]", e.getMessage());
+    }
+
+    public void testBadParam() throws IOException {
+        XContentParser parser = XContentType.JSON.xContent().createParser(
+                  "{\n"
+                + "  \"animal\": \"cat\",\n"
+                + "  \"vegetable\": 2,\n"
+                + "  \"a\": \"supercalifragilisticexpialidocious\"\n"
+                + "}");
+        ParsingException e = expectThrows(ParsingException.class, () -> HasRequiredArguments.PARSER.apply(parser, null));
+        assertEquals("[has_required_arguments] failed to parse field [a]", e.getMessage());
+        assertEquals(4, e.getLineNumber());
+        assertEquals("[a] must be less than 10 characters in length but was [supercalifragilisticexpialidocious]",
+                e.getCause().getMessage());
+    }
+
+    public void testBadParamBeforeObjectBuilt() throws IOException {
+        XContentParser parser = XContentType.JSON.xContent().createParser(
+                  "{\n"
+                + "  \"a\": \"supercalifragilisticexpialidocious\",\n"
+                + "  \"animal\": \"cat\"\n,"
+                + "  \"vegetable\": 2\n"
+                + "}");
+        ParsingException e = expectThrows(ParsingException.class, () -> HasRequiredArguments.PARSER.apply(parser, null));
+        assertEquals("[has_required_arguments] failed to parse field [vegetable]", e.getMessage());
+        assertEquals(4, e.getLineNumber());
+        e = (ParsingException) e.getCause();
+        assertEquals("failed to build [has_required_arguments] after last required field arrived", e.getMessage());
+        assertEquals(2, e.getLineNumber());
+        e = (ParsingException) e.getCause();
+        assertEquals("[has_required_arguments] failed to parse field [a]", e.getMessage());
+        assertEquals(2, e.getLineNumber());
+        assertEquals("[a] must be less than 10 characters in length but was [supercalifragilisticexpialidocious]",
+                e.getCause().getMessage());
+    }
+
+    public void testConstructorArgsMustBeConfigured() throws IOException {
+        class NoConstructorArgs {
+        }
+        ObjectSpooler<NoConstructorArgs, Void> spooler = new ObjectSpooler<>("constructor_args_required", (a) -> new NoConstructorArgs());
+        Exception e = expectThrows(IllegalStateException.class, () -> spooler.apply(XContentType.JSON.xContent().createParser("{}"), null));
+        assertEquals("[constructor_args_required] must configure at least on constructor argument. If it doens't have any it "
+                + "should use ObjectParser instead of ObjectSpooler. This is a bug in the spooler declaration.", e.getMessage());
+    }
+
+    /**
+     * Tests the non-constructor fields are only set on time.
+     */
+    public void testCalledOneTime() throws IOException {
+        class CalledOneTime {
```

Comment:
I think this should have a constructor argument to receive the value of the parsed field "yeah".